[C2][PyTorch][ONNX] End-to-end dynamic slicing with ONNX DynamicSlice experimental operator

jamesr66a opened the pull request:
Requires https://github.com/onnx/onnx/pull/1377

This PR makes it so that slices with dynamic boundary values can be exported from pytorch and run in caffe2 via ONNX.

apaszke commented on test/test_jit.py:
```diff
@@ -6486,6 +6486,25 @@ def foo(x):
 
         self.assertEqual(foo(input), input)
 
+    def test_export_dynamic_slice(self):
```
What's dynamic about this slice? That it doesn't have statically available ends?

jamesr66a replied:
Yeah, previously `starts`, `ends` and `axes` were required to be attributes in ONNX, this is dynamic in the sense that they can differ at runtime

apaszke reviewed (approved):
PyTorch parts LGTM. Didn't review C2 and ONNX changes

houseroad commented on caffe2/operators/slice_op.h:
```diff
@@ -198,22 +198,26 @@ bool SliceImpl(
       : Operator<Context>(operator_def, ws),
-        starts_(this->template GetRepeatedArgument<SIndex>("starts")),
-        ends_(this->template GetRepeatedArgument<SIndex>("ends")),
+        starts_(this->template GetRepeatedArgument<TIndex>("starts")),
```
This seems a breaking changes. We may have models in which `Slice` may take int32 tensor as inputs, with this change, they may break.

jamesr66a replied:
Note for anyone reading, this is actually not a breaking change since in the Proto the attributes are stored as int64 and this is doing a narrowing conversion under the hood

houseroad reviewed (commented):
I think the most desired feature in caffe2 is `axes` attribute. If we have that, we can make the conversion super easy. Also for the mismatched semantics of negative indexing, can we add a `numpy_compatible` attribute to caffe2, and use int64_max to indicate the end of the dimension?

houseroad reviewed (approved):
ONNX parts look good.

houseroad commented:
BTW, can we remove the dead code in the onnx=>caffe2 conversion?

Commit:
End-to-end dynamic slicing with ONNX DynamicSlice experimental operator (#11255)

Summary:
Requires https://github.com/onnx/onnx/pull/1377

This PR makes it so that slices with dynamic boundary values can be exported from pytorch and run in caffe2 via ONNX.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/11255

Differential Revision: D9790216

fbshipit-source-id: 83993872f66513c0d11cfe275b15aa66eba433da

jamesr66a commented:
@pytorchbot retest this please
